### PR TITLE
Construct DataVector from STL containers

### PR DESCRIPTION
**Pull request by nilsvu**
Various places in the code rolled their own solution to this.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/VectorImpl.hpp [hunk 1]
```diff
@@ -199,6 +199,18 @@ class VectorImpl
     std::fill(data(), data() + set_size, value);
   }
 
+  // Create from a copy of the given container
```
Comment (nilsdeppe) [inline src/DataStructures/VectorImpl.hpp]: triple slash

File: src/DataStructures/VectorImpl.hpp [hunk 2]
```diff
@@ -199,6 +199,18 @@ class VectorImpl
...
+  ///
+  /// \param container A container with a `value_type` that is the same as `T`,
+  /// e.g. `std::vector<T>` or `std::array<T>`.
+  template <typename Container, typename U = typename Container::value_type,
+            Requires<std::is_same_v<U, T>> = nullptr>
+  explicit VectorImpl(const Container& container)
```
Comment (nilsdeppe) [inline src/DataStructures/VectorImpl.hpp:210]: this is extremely generic and I'm worried people will do silly things that will be bad for performance. Can you please restrict it to vector & array explicitly?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by knelli2** [comment]
@nilsvu will look into this

File: src/DataStructures/VectorImpl.hpp [hunk 3]
```diff
@@ -199,6 +200,22 @@ class VectorImpl
     std::fill(data(), data() + set_size, value);
   }
 
+  /// Create from a copy of the given container
+  ///
+  /// \param container A container with a `value_type` that is the same as `T`.
+  /// Currently restricted to `std::vector<T>` and `std::array<T>`.
+  template <typename Container, typename U = typename Container::value_type,
```
Comment (nilsdeppe) [inline src/DataStructures/VectorImpl.hpp]: Why not inline the `Container::value_type` on the next line?

Comment (nilsvu) [reply src/DataStructures/VectorImpl.hpp]: done

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**